### PR TITLE
core: count references of notification events

### LINKED ISSUE #2482: Count notification data in the reference counter
Problematic wrt our refcounter implementation (but make some sense in general). neo-project/neo#2731

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #2482.

### CODE REVIEW

File: pkg/core/interop/runtime/engine.go [hunk 1]
```diff
@@ -74,7 +74,7 @@ func Notify(ic *interop.Context) error {
 		Name:       name,
 		Item:       stackitem.DeepCopy(stackitem.NewArray(args)).(*stackitem.Array),
 	}
-	ic.Notifications = append(ic.Notifications, ne)
+	ic.Notifications = append(ic.Notifications, ne) // Do not use ic.VM.AddRef, because `args` item was counted earlier.
```
Comment (roman-khimov) [inline pkg/core/interop/runtime/engine.go:77]: But we make a `DeepCopy` of it, so technically it's a different element. Also, `elem` is popped off the stack already.

File: pkg/core/native/designate.go [hunk 2]
```diff
@@ -386,14 +386,16 @@ func (s *Designate) DesignateAsRole(ic *interop.Context, r noderoles.Role, pubs
...
 	ic.Notifications = append(ic.Notifications, state.NotificationEvent{
 		ScriptHash: s.Hash,
 		Name:       DesignationEventName,
-		Item: stackitem.NewArray([]stackitem.Item{
-			stackitem.NewBigInteger(big.NewInt(int64(r))),
-			stackitem.NewBigInteger(big.NewInt(int64(ic.Block.Index))),
-		}),
+		Item:       itm,
 	})
+	ic.VM.AddRef(itm)
```
Comment (roman-khimov) [inline pkg/core/native/designate.go:398]: Maybe we better add some `ic.AddNotification(hash, name, item)` that will handle this refcounting internally (as well as `state.NotificationEvent` creation)?

File: pkg/core/interop/runtime/engine.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/core/interop/runtime/engine.go:77]: You're right, in C# it is done a bit earlier during `args` conversion to stackitem:
https://github.com/neo-project/neo/blob/ed68274e7dec2622468d3bc67df6e0f2913d903b/src/neo/SmartContract/ApplicationEngine.cs#L404

File: pkg/core/native/designate.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/core/native/designate.go]: Yes, fixed.

File: pkg/core/interop/runtime/engine.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/core/interop/runtime/engine.go]: Fixed.

**Comment by roman-khimov** [comment]
Looks like I've misinterpreted the original change, see https://github.com/neo-project/neo/pull/2729#discussion_r874303243. We don't have this issue at all, so this PR isn't needed.

**Comment by roman-khimov** [comment]
To me, yes, it looks like we're adding something to the refcounter.

**Comment by AnnaShaleva** [comment]
I got it, it adds the array to the refcounter and updates `ObjectReferences`, but it doesn't modify `StackReferences`. So the emitted notification is not on the stack, and thus, it shouldn't be taken into account while checking references after subsequent VM instruction handling.

**Pull request closed without merging**